bsp/Src/ADC.c

```diff
@@ -0,0 +1,163 @@
+            hadc1->Init = init;
+            if (HAL_ADC_Init(hadc1) != HAL_OK) return ADC_INIT_FAIL;
+            break;
+        case 2:
+            __HAL_RCC_ADC2_CLK_ENABLE();
+            hadc2->Init = init;
+            if (HAL_ADC_Init(hadc2) != HAL_OK) return ADC_INIT_FAIL;
+            break;
+        case 3:
+            __HAL_RCC_ADC3_CLK_ENABLE();
+            hadc3->Init = init;
+            if (HAL_ADC_Init(hadc3) != HAL_OK) return ADC_INIT_FAIL;
+            break;
+        default:
+            // Couldn't initalize any ADC
+            return ADC_INIT_FAIL;
+            break;
+    }
+
+    // Configure NVIC
+    HAL_NVIC_SetPriority(ADC_IRQn, 5, 0);
+    HAL_NVIC_EnableIRQ(ADC_IRQn);
+    
+    return ADC_OK;
+}
+
+
+adc_status_t ADC_OneShotRead(uint32_t channel, uint32_t samplingTime, uint8_t adcNum, QueueHandle_t* q) {
+    if (xQueueIsQueueFullFromISR(*q)) {
+        return ADC_QUEUE_FULL;
+    }
+
+    // Reset channel config for only one channel
+    ADC_ChannelConfTypeDef sConfig = {
+        .Channel = channel,
+        .Rank = 1,
+        .SamplingTime = samplingTime
+    }; 
+    
+    // --- Config channel
+    // --- Start ADC Interrupt
+    switch (adcNum) {
+        case 1: 
+            adc1_q = q;
+
+            if (HAL_ADC_ConfigChannel(hadc1, &sConfig) != HAL_OK) {
+                return ADC_CHANNEL_CONFIG_FAIL;
+            }
+
+            if (HAL_ADC_Start_IT(hadc1) != HAL_OK) {
+                return ADC_INTERRUPT_FAIL;
+            }
+
+            break;
+        case 2:
+            adc2_q = q;
+
+            if (HAL_ADC_ConfigChannel(hadc2, &sConfig) != HAL_OK) {
+                return ADC_CHANNEL_CONFIG_FAIL;
+            }
+
+            if (HAL_ADC_Start_IT(hadc2) != HAL_OK) {
+                return ADC_INTERRUPT_FAIL;
+            }
+
+            break;
+        case 3: 
+            adc3_q = q;
+            
+            if (HAL_ADC_ConfigChannel(hadc3, &sConfig) != HAL_OK) {
+                return ADC_CHANNEL_CONFIG_FAIL;
+            }
+            
+            if (HAL_ADC_Start_IT(hadc3) != HAL_OK) {
+                return ADC_INTERRUPT_FAIL;
+            }
+
+            break;
+        default:
+            // ADC out of range
+            return ADC_CHANNEL_CONFIG_FAIL;
+    }
+    
+    return ADC_OK; 
+}
+
+
+void HAL_ADC_ConvCpltCallback(ADC_HandleTypeDef *h) {
+    /*
+    Interrupt triggers this callback when the conversion is complete
+    */ 
+   BaseType_t higherPriorityTaskWoken = pdFALSE;
+   int rawVal;
+
+    if (h == hadc1) {
+        rawVal = HAL_ADC_GetValue(hadc1);
+
+        xQueueSendFromISR(*adc1_q, &rawVal, &higherPriorityTaskWoken);
+
+    } else if (h == hadc2) {
+        rawVal = HAL_ADC_GetValue(hadc1);
+        xQueueSendFromISR(*adc2_q, &rawVal, &higherPriorityTaskWoken);
+    
+    } else if (h == hadc3) {
+        rawVal = HAL_ADC_GetValue(hadc1);
+        xQueueSendFromISR(*adc3_q, &rawVal, &higherPriorityTaskWoken);
+        
+    }
```

Comment:
This is good. U can also use the #ifdefs to conditionally compile these in based on the target, so it won't even check for hadc3 if there's no ADC3 on the target platform (similar to how u did it on lines 4 thru 17

Reply:
Ye I can add ifdefs in between the if statements.